polar-core/src/rewrites.rs

```diff
@@ -184,14 +184,23 @@ impl<'kb> Folder for Rewriter<'kb> {
                         //
                         // We prepend when the rewritten variable needs to be bound before it is
                         // used.
-                        if only_dots(&rewrites)
-                            && arg_operator.map_or(false, |o| o == Operator::Unify)
-                        {
-                            for rewrite in rewrites {
+                        use Operator::*;
+                        match (only_dots(&rewrites), arg_operator) {
+                            (true, Some(Unify)) => for rewrite in rewrites {
                                 and_append(&mut arg, rewrite);
+                            },
+                            (true, Some(ForAll)) => {
+                                let Operation { args, .. } = arg.value().as_expression().unwrap().clone();
+                                rewrites.insert(0, args[1].clone());
+                                arg.replace_value(Value::Expression(Operation {
+                                    operator: ForAll,
+                                    args: vec![args[0].clone(), term!(Operation {
+                                        operator: And,
+                                        args: rewrites
+                                    })],
+                                }));
                             }
-                        } else {
-                            for rewrite in rewrites.drain(..).rev() {
+                            _ => for rewrite in rewrites.drain(..).rev() {
```

Comment:
I think it would be helpful to add a rewrite test for the `ForAll` case. I've looked at those in the past when working on the rewriter both to figure out how things are rewritten when I forget (as a form of documentation), and to make it easier to refactor / make changes to the code rather than running an integration test.

what do you think?

Reply:
Hmm... I'm somewhat not in favor because I think we should be testing for meaning not format here. In the simplifier it would be useful because we kind of use that as an interchange format, but here all that matters is if the Polar program runs correctly. We don't really care about, or need to document whether it's `(and (and (and a b) c) d)` or `(and a b c d)` at this point.

Reply:
Actually never mind, you were right, doing this helped improve the fix a lot, thank you 😄